core/shared/src/main/scala/fs2/Chunk.scala

```diff
@@ -703,6 +828,144 @@ object Chunk {
       Doubles(values, 0, values.length)
   }
 
+  /** Concatenates the specified sequence of chunks in to a single chunk, avoiding boxing. */
+  def concat[A](chunks: Seq[Chunk[A]]): Chunk[A] =
+    if (chunks.isEmpty) {
+      Chunk.empty
+    } else if (chunks.forall(c => c.knownElementType[Boolean] || c.forall(_.isInstanceOf[Boolean]))) {
+      concatBooleans(chunks.asInstanceOf[Seq[Chunk[Boolean]]]).asInstanceOf[Chunk[A]]
+    } else if (chunks.forall(c => c.knownElementType[Byte] || c.forall(_.isInstanceOf[Byte]))) {
+      concatBytes(chunks.asInstanceOf[Seq[Chunk[Byte]]]).asInstanceOf[Chunk[A]]
```

Comment:
This doesn't solve the `ByteVectorChunk` problem. Or `ByteBuffer`. Both of them will be paged out to `Array[Byte]` and then run through `concatBytes`.

Reply:
Both `ByteVectorChunk` and `ByteBuffer` implement `KnownElementType[Byte]` and both have reasonably efficient `copyToArray` implementations (`ByteBuffer` more so than `ByteVector`). What am I missing?

Reply:
I guess there really isn't an efficient way to do `ByteBuffer` append intrinsically. One of the reasons why `ByteVector` is nicer. ;-)

With `ByteVector` though, you don't *want* to copy out to an array: you want to use the intrinsic append. This is actually really really important, because if you have a `ByteVector` which is wrapped around a direct `ByteBuffer`, you can append it without ever paging those bytes into heap memory. If you use `copyToArray`, you're sort of screwed.

Reply:
Right, agreed. It would be much nicer to use stuff like `ByteVector.++` if most chunks are `ByteVector` backed, though that's not *much* different than using `Catenable`. I think we had polymorphic dispatch based on type of first chunk at some point, but maybe that never made it passed the discussion phase. That's definitely not what ended up in 0.9 final though.

Whatever we do, it has to account for things like `Chunk.empty` and `Chunk.singleton` being intermixed with the high density chunk types. Also should probably handle `Chunk.buffer` as well since that can show up when dealing with small literal chunks (e.g. `Chunk(1, 2, 3)`).

Reply:
> Whatever we do, it has to account for things like Chunk.empty and Chunk.singleton being intermixed with the high density chunk types. Also should probably handle Chunk.buffer as well since that can show up when dealing with small literal chunks (e.g. Chunk(1, 2, 3)).

I'll give it some thought. I think this is doable and not too difficult to PR.